pallets/settlement/src/benchmarking.rs

```diff
@@ -216,33 +231,345 @@ fn emulate_add_instruction<T: Trait>(
+            let transacted_amount = 500 * POLY;
+            let sender_portfolio = generate_portfolio::<T>("", index, 500, sender);
+            let receiver_portfolio = generate_portfolio::<T>("", index, 500, receiver);
+            let _ = fund_portfolio::<T>(&sender_portfolio, ticker, transacted_amount.into());
+            portfolios.push(sender_portfolio);
+            default_portfolio.push(receiver_portfolio);
+            legs.push(Leg {
+                from: sender_portfolio,
+                to: receiver_portfolio,
+                asset: *ticker,
+                amount: transacted_amount.into(),
+            })
+        };
+    emulate_portfolios(
+        from_did,
+        to_did,
+        sender_portfolios,
+        &from_ticker,
+        receiver_portfolios,
+    );
+    emulate_portfolios(
+        to_did,
+        from_did,
+        receiver_portfolios,
+        &to_ticker,
+        sender_portfolios,
+    );
+}
+
+// Generate signature.
+fn get_encoded_signature<T: Trait>(signer: &User<T>, msg: &Receipt<T::Balance>) -> Vec<u8> {
+    let raw_signature: [u8; 64] = signer.sign(&msg.encode()).expect("Data cannot be signed").0;
+    let encoded = MultiSignature::from(Signature::from_raw(raw_signature)).encode();
+    encoded
+}
+
+// Add investor uniqueness claim directly in the storage.
+fn add_investor_uniqueness_claim<T: Trait>(did: IdentityId, ticker: Ticker) {
+    identity::Module::<T>::base_add_claim(
+        did,
+        Claim::InvestorUniqueness(
+            Scope::Ticker(ticker),
+            did,
+            CddId::new(did, InvestorUid::from(did.to_bytes())),
+        ),
+        did,
+        None,
+    );
+    let current_balance = <pallet_asset::Module<T>>::balance_of(ticker, did);
+    <AggregateBalance<T>>::insert(ticker, &did, current_balance);
+    <BalanceOfAtScope<T>>::insert(did, did, current_balance);
+}
+
+fn add_trusted_issuer<T: Trait>(
+    origin: RawOrigin<T::AccountId>,
+    ticker: Ticker,
+    issuer: TrustedIssuer,
+) {
+    pallet_compliance_manager::Module::<T>::add_default_trusted_claim_issuer(
+        origin.into(),
+        ticker,
+        issuer,
+    )
+    .expect("Default trusted claim issuer cannot be added");
+}
+
+pub fn create_condition<T: Trait>(
+    c_count: u32,
+    t_count: u32,
+    ticker: Ticker,
+    token_owner: RawOrigin<T::AccountId>,
+) -> Vec<Condition> {
+    let condition_type = get_condition_type::<T>(&c_count, Scope::Ticker(ticker));
+    let trusted_issuers = make_issuers::<T>(t_count);
+    trusted_issuers.clone().into_iter().for_each(|issuer| {
+        let t_issuer = <pallet_compliance_manager::Module<T>>::trusted_claim_issuer(ticker);
+        if !t_issuer.contains(&issuer) {
+            add_trusted_issuer::<T>(token_owner.clone().into(), ticker, issuer);
+        }
+    });
+    vec![Condition::new(condition_type, trusted_issuers)]
+}
+
+pub fn get_condition_type<T: Trait>(condition_count: &u32, scope: Scope) -> ConditionType {
```

Comment:
I don't see the need for this. Instead, let's maybe just use

```
fn get_condition(complexity: u8) -> ConditionType {
    let conditions = Vec::with_capacity(complexity);
    for i in 0..complexity {
        conditions.push(Claim::Jurisdiction(CountryCode::AF, Scope::custom(i.into())));
    } 
    ConditionType::IsNoneOf(conditions)
}
```
this will reliably produce the same result while still avoiding any cached entries.

Reply:
I don't think so, Because complexity doesn't give you the worst case. It can be a worst-case for setting the rules but for transfer, it will never give you the worst case. Because It doesn't cover the case where every compliance rules get checked iteratively and got matched in last, to make it O(n) complexity where n is the no. of condition rules applied for a given asset transfer.

Reply:
> Because It doesn't cover the case where every compliance rules get checked iteratively and got matched in last, to make it O(n) complexity where n is the no. of condition rules applied for a given asset transfer.

Complexity is mostly dependent on the number of conditions, not rules. For benchmarks, use a single rule with multiple conditions as I suggested above. Your current approach is overly complex and does not give reliable results.

Reply:
I argue over it. The worst condition can only be achieved when sender and receiver conditions are met at the last requirement, but the complexity never allowed you to achieve the worst condition. It will always be the combination of no. of conditions in a requirement and no. of the requirement itself. I think more or less we are saying things but other way around.
FYI it never gives you the cached entries.

Reply:
> I argue over it. The worst condition can only be achieved when sender and receiver conditions are met at the last requirement, but the complexity never allowed you to achieve the worst condition. It will always be the combination of no. of conditions in a requirement and no. of the requirement itself. I think more or less we are saying things but other way around.

The complexity is n*m. This does not mean you need to bench both n, m separately. You can just set n as a constant 1 and vary m. The advantage of setting n as 1 here is that it will always be the worse case since all conditions are processed in any given rule for it to pass. This will simplify everything by a lot.

> FYI it never gives you the cached entries.

Rules that fetch the same claim will give you cached entries.

Reply:
Now rules are changed as the scope is different in every claim. Anyway, I will make that change to simplify things.